server/src/main/java/org/elasticsearch/search/aggregations/timeseries/TimeSeriesIndexSearcher.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.aggregations.timeseries;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.SortedNumericDocValues;
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.LeafCollector;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.Bits;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.PriorityQueue;
+import org.elasticsearch.cluster.metadata.DataStream;
+import org.elasticsearch.index.mapper.TimeSeriesIdFieldMapper;
+
+import java.io.IOException;
+
+/**
+ * An IndexSearcher wrapper that executes the searches in time-series indices by traversing them by tsid and timestamp
+ * TODO: Convert it to use index sort instead of hard-coded tsid and timestamp values
+ */
+public class TimeSeriesIndexSearcher {
```

Comment:
Just so I'm clear about why this is required: we want to collect documents for a given `tsid` in order of strictly increasing timestamp? And so just going segment-by-segment won't work because individual metrics may be interleaved across multiple segments? I think there is probably an easier way to do this, but I need to think more about it!

Reply:
We essentially need to walk the records in the in-index order (tsid, timestamp) across all segments. Individual metrics defined by tsid will be indeed interleaving across multiple segments.

Reply:
Algorithm-wise is it mostly rate and different moving average-like function. But it also allows us to see when we collected all records from a particular bucket, which enables a whole range of other optimizations allowing us to significantly reduce the amount of information that a data node has to send to a coordinating node since we no longer need to reduce the bucket there.